dataset.py

```diff
@@ -139,6 +182,226 @@ def _partition_data_random(sample_data: pd.DataFrame,
+  assert(
+    min_size <= len(coordinates)
+  ), (f"You need {min_size} aka {int(len(coordinates) * strategy.train_fraction)} train + "
+    f"{int(len(coordinates) * strategy.validation_fraction)} validation + "
+    f"{int(len(coordinates) * strategy.test_fraction)} test samples but you have a sample of {len(coordinates)}")
+  
+  return coordinates
+
+def _maybe_partition_furthest_points(
+  sample_data: pd.DataFrame,
+  furthest_coordinates: list[list[float]],
+  strategy: FurthestPointsPartitionStrategy) -> PartitionedDataset:
+  '''
+  Returns a PartitionedDataset that has a valid train/validation/test split
+  that has:
+  - Non-overlapping splits
+  - Splits which polygon areas have less than 0.1 area
+  It attempts to generate these partitions randomly for at most strategy.max_attempts.
+  If unsuccesful, returns None.
+  '''
+  partitioned_dataset = None
+  attempts = 0
+  while (attempts <= strategy.max_attempts and
+         partitioned_dataset is None):
+    random.seed(strategy.random_seed)
+    random.shuffle(furthest_coordinates)
+    # We need permutations of 3 numbers that correspond to train, validation and test.
+    coord_permutations = list(permutations(furthest_coordinates, 3))
+    random.shuffle(coord_permutations)
+    # Pick one random permutation.
+    sampled_permutation = coord_permutations[0]
+
+    partitioned_dataset = _partition_by_nearest_neighbors(
+      sample_data=sample_data,
+      strategy=strategy,
+      train_coord=sampled_permutation[0],
+      validation_coord=sampled_permutation[1],
+      test_coord=sampled_permutation[2]
+    )
+    train_polygon = _polygon(partitioned_dataset.train[
+      [_LONGITUDE_COLUMN_NAME, _LATITUDE_COLUMN_NAME]
+    ].values)
+    validation_polygon = _polygon(partitioned_dataset.validation[
+      [_LONGITUDE_COLUMN_NAME, _LATITUDE_COLUMN_NAME]
+    ].values)
+    test_polygon = _polygon(partitioned_dataset.test[
+      [_LONGITUDE_COLUMN_NAME, _LATITUDE_COLUMN_NAME]
+    ].values)
+    are_valid_polygons = _valid_polygons(
+      train_polygon=train_polygon, validation_polygon=validation_polygon,
+      test_polygon=test_polygon)
+    if (train_polygon is None or
+        validation_polygon is None or
+        test_polygon is None or
+        not are_valid_polygons):
+      partitioned_dataset = None
+    attempts += 1
+  return partitioned_dataset
+
+def _partition_data_furthest_points(
+  sample_data: pd.DataFrame,
+  strategy: FurthestPointsPartitionStrategy) -> PartitionedDataset:
+  coordinates = _shuffled_unique_coordinates(sample_data, strategy)
+  
+  # Compute centroid of the coordinates
+  centroid = [sum([c[0] for c in coordinates])/len(coordinates),
+              sum([c[1] for c in coordinates])/len(coordinates)]
+  
+  # Sort coordinates from furthest to closest to the centroid.
+  distances_to_centroid = []
+  for coord in coordinates:
+    distances_to_centroid.append((
+      math.dist(centroid, coord), (coord[0], coord[1])))
```

Comment:
It might be good to specify in the externally-facing docstring that the specific distance we use is Euclidean.

Reply:
Done.